docs/upgrading_to_bigquery_v4.0.md

```diff
@@ -0,0 +1,98 @@
+# Upgrading to BigQuery v4.0
+
+The v4.0 release of BigQuery is a backwards incompatible release.
+- The supported provider has been updated to v3.0.
+- Add support for setting custom access. The default has been changed to be more
+  secure.
+- The UDFS submodule has been isolated from the main module.
+- Trimmed outputs and made their names more consistent.
+
+## Access
+
+The major logic change in this release is a new default for the `access` field.
+
+Datasets are created with the following
+[default](https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets):
+
+
+> If unspecified at dataset creation time, BigQuery adds default dataset access for the following entities: access.specialGroup: projectReaders; access.role: READER; access.specialGroup: projectWriters; access.role: WRITER; access.specialGroup: projectOwners; access.role: OWNER; access.userByEmail: [dataset creator email]; access.role: OWNER;
+
+The problematic default is
+`access.userByEmail: [dataset creator email]; access.role: OWNER`.
+This will subtly grant owner access to the creating user and leave that access
+on, even if the acess to the project, etc is revoked.
+
+The new default will only set
+`specialGroup: projectOwners; access.role: roles/bigquery.admin`. This is done
+because the BigQuery API requires at least one owner on the dataset, and this is
+the safest and most generic owner that can be granted.
+
+Do note that standard GCP IAM logic still apply to bigquery datasets.
+For example, setting an IAM role at the project level will inherit to all
+datasets within the project.
+
+## Migration Instructions
+
+1. Upgrade version
+```diff
+module "bigquery" {
+  source  = "terraform-google-modules/bigquery/google"
+-  version = "~> 3.0"
++  version = "~> 4.0"
+....
+}
+```
+
+2. Set access if you wish to preserve existing behaviour
+```diff
+module "bigquery" {
+  source  = "terraform-google-modules/bigquery/google"
++ access {
```

Comment:
Two issues:

1. The syntax needs to be access = [{role: ...}, {role: ...}]
2. Could we also point out the option of not setting access at all. ie `access = []` which would leave the access setting untouched.

Reply:
1. Ops, was thinking in the resource syntax. Fixed.

2. IIUC this should not be the case. Not setting the access should set the following default:
```
  [{
    role          = "roles/bigquery.dataOwner"
    special_group = "projectOwners"
  }]
```

Which is not the same as manually setting access = [] or access = null. And this is a desired behaviour so that a more secure access is set on existing dataset. To leave access untouched/not show a diff the user would have to set access as described here. Am I missing something?

Reply:
Right, not setting access at all is different from setting `access = []`. If users don't add `access`, we will update it to match the new (secure) default.

But if users want to retain their existing config they can set `access = []` which will override the default and *no* access blocks will be [created](https://github.com/terraform-google-modules/terraform-google-bigquery/blob/master/main.tf#L31) thus preserving the existing behavior without having to list out every role. Does that make sense?

NB: I haven't tested this. Please give it a try before we include it in the migration guide.

Reply:
Ok, I get what you're suggesting now.

Here are the results:

* access unset:
```
      + access {
          + role          = "roles/bigquery.dataOwner"
          + special_group = "projectOwners"
        }
```

* access set to []

```
 access {
          + domain         = (known after apply)
          + group_by_email = (known after apply)
          + role           = (known after apply)
          + special_group  = (known after apply)
          + user_by_email  = (known after apply)

          + view {
              + dataset_id = (known after apply)
              + project_id = (known after apply)
              + table_id   = (known after apply)
            }
        }
```

This is actually unexpected behaviour to me as I would expect [] to try and set access to an empty list. Reason being it doesn't set the value in the API call if it's an empty slice: 

https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_big_query_dataset.go#L313

Given the above, I simplified the migration instructions to just say set access to [] as you suggested.